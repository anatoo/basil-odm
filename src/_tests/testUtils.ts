import * as ts from "typescript";
import { prettier } from "../utils";

// TypeScriptのコードを整形して、比較可能な文字列に正規化する
export function format(code: string): string {
  const sourceFile = ts.createSourceFile("", code, ts.ScriptTarget.ES2015);
  const printer = ts.createPrinter();

  return prettier(printer.printFile(sourceFile));
}
