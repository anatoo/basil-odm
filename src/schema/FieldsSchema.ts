import { getSchemaFragment } from './utils';
import { literal } from './literal';
import { Document, Entity, ObjectSchemaSource, SchemaFragment, SchemaFragmentAggregate, SchemaLike } from './types';
import { Field, FieldsSchemaRoot } from './astTypes';
import { createDocument, createEntity } from '../extract';
import { generateBsonSchema } from '../generateBsonSchema';
import { getSchemaFragmentSymbol, optionalPropertyFlag, schemaFragmentFrag } from './symbols';

(String.prototype as unknown as SchemaFragmentAggregate)[getSchemaFragmentSymbol] = function (this: string) {
  return literal(this);
};

(Number.prototype as unknown as SchemaFragmentAggregate)[getSchemaFragmentSymbol] = function (this: number) {
  return literal(this);
};

(Boolean.prototype as unknown as SchemaFragmentAggregate)[getSchemaFragmentSymbol] = function (this: boolean) {
  return literal(this);
};

(Array.prototype as unknown as SchemaFragmentAggregate)[getSchemaFragmentSymbol] = function (this: Array<unknown>) {
  if (this.length === 0) {
    throw Error('You must pass one element at least into array.');
  }

  if (this.length > 1) {
    throw Error('You must pass one element into array on building schema.');
  }

  return arrayOf(this[0] as SchemaLike);
};

(Object.prototype as unknown as SchemaFragmentAggregate)[getSchemaFragmentSymbol] = function () {
  return shape(this as unknown as ObjectSchemaSource);
};

// entityが余分なプロパティを持っていても許す
export function shape<T extends ObjectSchemaSource>(source: T): SchemaFragment {
  const object: Record<string, SchemaFragment> = {};

  Object.keys(source).forEach((key: string) => {
    object[key] = getSchemaFragment(source[key]);
  });

  return {
    [schemaFragmentFrag]: true,

    buildASTNode(): FieldsSchemaRoot {
      const props: Record<string, Field> = {};

      for (const [key, value] of Object.entries(object)) {
        props[key] = {
          kind: 'field',
          isOptional: !!value[optionalPropertyFlag],
          node: value.buildASTNode(),
        };
      }

      return {
        kind: 'object',
        props,
        allowAdditionalProps: Object.keys(props).length === 0,
      };
    },
  };
}

export function arrayOf<T extends SchemaLike>(item: T): SchemaFragment {
  const fragment = getSchemaFragment(item);

  return {
    [schemaFragmentFrag]: true,

    buildASTNode() {
      return {
        kind: 'array',
        item: fragment.buildASTNode(),
      };
    },
  };
}

class Union implements SchemaFragment {
  schemaFragments: SchemaFragment[];

  [schemaFragmentFrag] = true as const;

  constructor(schemas: SchemaLike[]) {
    this.schemaFragments = schemas.map((schema) => getSchemaFragment(schema));
  }

  buildASTNode() {
    return {
      kind: 'union' as const,
      items: this.schemaFragments.map((fragment) => fragment.buildASTNode()),
    };
  }
}

export function createFieldsSchema(source: ObjectSchemaSource) {
  const fragment = shape(source);
  const node = fragment.buildASTNode();

  if (node.kind !== 'object') {
    throw Error('Invalid state. This is bug.');
  }

  return new FieldsSchema(node);
}

export class FieldsSchema {
  _schemaRoot: FieldsSchemaRoot;

  constructor(schemaRoot: FieldsSchemaRoot) {
    this._schemaRoot = schemaRoot;
  }

  getSchemaAST(): FieldsSchemaRoot {
    return this._schemaRoot;
  }

  decode(entity: Entity): Document {
    return createDocument(entity, this.getSchemaAST()) as Document;
  }

  encode(document: Document, source: Record<string, unknown> = {}): Entity {
    return createEntity(source, document, this.getSchemaAST()) as Entity;
  }

  generateBsonSchema(): Record<string, unknown> {
    return generateBsonSchema(this.getSchemaAST());
  }
}

export function arrayOfShape<T extends ObjectSchemaSource>(object: T): SchemaFragment {
  return arrayOf(shape(object));
}

export function union(...schemas: SchemaLike[]): SchemaFragment {
  return new Union(schemas);
}
