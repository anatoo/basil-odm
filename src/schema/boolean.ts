import { SchemaFragment } from './types';
import { schemaFragmentFrag } from './symbols';

function originalBoolean(): SchemaFragment {
  return {
    [schemaFragmentFrag]: true,

    buildASTNode() {
      return {
        kind: 'boolean',
      };
    },
  };
}

export const boolean: typeof originalBoolean & SchemaFragment = Object.assign(originalBoolean, originalBoolean());
