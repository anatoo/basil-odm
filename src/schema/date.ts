import { SchemaFragment } from './types';
import { schemaFragmentFrag } from './symbols';

function originalDate(): SchemaFragment<Date> {
  return {
    [schemaFragmentFrag]: true,

    buildASTNode() {
      return { kind: 'date' };
    },
  };
}

export const date: typeof originalDate & SchemaFragment<Date> = Object.assign(originalDate, originalDate());
