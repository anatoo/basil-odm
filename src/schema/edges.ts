import { HasOne, HasMany } from './edgeTypes';

export function hasOne(props: Omit<HasOne, 'type'>): HasOne {
  return {
    type: 'hasOne',
    ...props,
  };
}

export function hasMany(props: Omit<HasMany, 'type'>): HasMany {
  return {
    type: 'hasMany',
    ...props,
  };
}
