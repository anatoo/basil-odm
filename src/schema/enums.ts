import { schemaFragmentFrag } from './symbols';
import { SchemaFragment } from './types';
import { inspect } from 'util';

export function enums(props: { values: Record<string, Value> | Record<number, Value> }): SchemaFragment {
  const { values } = props;

  return new Enum({
    values: Array.isArray(values)
      ? values.reduce((record, item, index) => {
          record['' + index] = item;
          return record;
        }, {})
      : values,
  });
}

type Value = string | boolean | number | null;

interface Props {
  values: Record<string, Value>;
}

class Enum implements SchemaFragment {
  values: Record<string, Value>;

  [schemaFragmentFrag]: true = true as const;

  constructor(props: Props) {
    const { values } = props;

    Object.values(values).forEach((value) => {
      if (typeof value !== 'string' && typeof value !== 'boolean' && typeof value !== 'number' && value !== null) {
        throw Error(`Enum values must be a string, boolean, number, or null. Got "${inspect(value)}".`);
      }
    });

    this.values = values;
  }

  buildASTNode() {
    return {
      kind: 'enum' as const,
      values: this.values,
    };
  }
}
