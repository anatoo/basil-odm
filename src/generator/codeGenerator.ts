import { FieldsSchemaRoot, Enum } from '../schema/astTypes';
import { CollectionDef, DefinedSchema } from '../types';
import { writeFileSync } from 'fs';
import { prettier } from '../utils';
import { aggregateEnums, generateDefaultValue, generateType, getDefaultEntityName } from './utils';
import { inspect } from 'util';

export const header = `/**
 * This file was generated by Basil ODM
 * Do not make changes to this file directly
 */
/* eslint-disable */
import * as mongodb from 'mongodb';
import * as basil from 'basil-odm';

`;

export function generateEnumsCode(collections: CollectionDef[]): string {
  let code = '';

  let enums = collections.flatMap((collection) => {
    const root = collection.schema.getSchemaAST();
    return aggregateEnums(root);
  });

  // remove duplicated
  const enumsMap = new Map<string, Enum>();
  for (const node of enums) {
    if (typeof node.name === 'string') {
      if (enumsMap.has(node.name)) {
        const storedNode = enumsMap.get(node.name) as Enum;
        if (storedNode !== node && JSON.stringify(storedNode) !== JSON.stringify(node)) {
          throw Error(`These enums node has same name but different object: ${inspect(node)} ${inspect(storedNode)}`);
        }
      } else {
        enumsMap.set(node.name, node);
      }
    }
  }

  enums = [...enumsMap.values(), ...enums.filter((node) => typeof node.name !== 'string')];

  for (const node of enums) {
    if (node.name) {
      code += `export const ${node.name} = {`;
      for (const [key, value] of Object.entries(node.values)) {
        code += `${JSON.stringify(key)}: ${JSON.stringify(value)},`;
      }
      code += `} as const;\n`;

      code += `export type ${node.name} = typeof ${node.name}[keyof typeof ${node.name}];\n\n`;
    }
  }

  return code;
}

export function generateTypeScriptFile(collections: CollectionDef[]): string {
  return header + generateCollectionDefs(collections) + generateDocumentTypes(collections) + generateEnumsCode(collections);
}

export function generateCollectionDefs(collections: CollectionDef[]): string {
  let code = '';

  code += `const $defs: Record<string, basil.EntityMeta<Record<string, unknown>>> = {`;
  for (const collection of collections) {
    code += `
${JSON.stringify(collection.collectionName)}: {
  collectionName: ${JSON.stringify(collection.collectionName)},
  schema: new basil.FieldsSchema(${JSON.stringify(collection.schema.getSchemaAST())}),
  indexes: ${JSON.stringify(collection.indexes)},
},`;
  }
  code += `};\n\n`;

  return code;
}

export function generateDocumentTypes(collections: CollectionDef[]): string {
  let code = ``;

  for (const collection of collections) {
    const ast = collection.schema.getSchemaAST();
    const name = collection.entityName ?? getDefaultEntityName(collection.collectionName);

    code += `export class ${name} extends basil.Base {
constructor(params?: Partial<${name}>) {
  super();
  Object.assign(this, params);
}
static getCollection() {
  return {
    collectionName: ${JSON.stringify(collection.collectionName)},
    schema: $defs[${JSON.stringify(collection.collectionName)}].schema,
    indexes: $defs[${JSON.stringify(collection.collectionName)}].indexes,
  };
}`;

    for (const [name, field] of Object.entries(ast.props)) {
      code += `${name}${field.isOptional ? '?' : ''}: ${generateType(field.node)}`;
      if (!field.isOptional) {
        code += '= ' + generateDefaultValue(field.node);
      }
      code += ';\n';
    }

    code += '}\n\n';
  }

  return code;
}

export function generateTypeFromSchema(root: FieldsSchemaRoot) {
  return generateType(root);
}

interface GenerateCodeParams {
  outputFile: string;
  schema: DefinedSchema;
}

export function generateCode(params: GenerateCodeParams) {
  const { schema, outputFile } = params;
  const collections = Object.values(schema);
  writeFileSync(outputFile, prettier(generateTypeScriptFile(collections)));
}
