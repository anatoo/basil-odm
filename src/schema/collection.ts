import { CollectionDef, Index, CollectionOptions } from '../types';
import { createFieldsSchema } from './FieldsSchema';
import { ObjectSchemaSource } from './types';
import { Edge } from './edgeTypes';
import { getDefaultEntityName } from '../generator/utils';
interface Props {
  collectionName: string;
  fields?: ObjectSchemaSource;
  indexes?: Index[];
  entityName?: string;
  options?: CollectionOptions;
  edges?: Record<string, Edge>;
}

export function collection(props: Props): CollectionDef {
  return {
    collectionName: props.collectionName,
    entityName: props.entityName || getDefaultEntityName(props.collectionName),
    fields: createFieldsSchema(props.fields || {}),
    indexes: props.indexes || [],
    options: props.options,
    edges: props.edges,
  };
}
