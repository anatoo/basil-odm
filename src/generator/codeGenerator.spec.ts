import { enums } from '../schema/enums';
import { generateDocumentTypes } from './codeGenerator';
import { collection } from '../schema/collection';
import { objectId } from '../schema/objectId';
import { date } from '../schema/date';
import { boolean } from '../schema/boolean';
import { string } from '../schema/string';
import { record } from '../schema/record';
import { hasOne } from '../schema/edges';
import { CollectionDef } from '../types';
import { format } from '../testUtils';
import { shape } from '../schema/FieldsSchema';

const table: {
  collections: CollectionDef[];
}[] = [
  {
    collections: [
      collection({
        collectionName: 'docs',
        fields: {
          _id: objectId(),
          createdAt: date(),
          flag: boolean(),
          status: enums({ values: ['created', 'deleted'] }),
          name: string(),
          sub: shape({
            name: string(),
          }),
          record: record(string),
        },
      }),
    ],
  },
  {
    collections: [
      collection({
        collectionName: 'docs',
        fields: {
          _id: objectId(),
          userId: objectId(),
        },
        edges: {
          user: hasOne({
            collection: 'users',
            referenceField: 'userId',
          }),
        },
      }),
      collection({
        collectionName: 'users',
        fields: {
          _id: objectId(),
        },
      }),
    ],
  },
];

test.each(table)('generateDocumentTypes() #%#', ({ collections }) => {
  const code = generateDocumentTypes(collections);
  expect(format(code)).toMatchSnapshot();
});
