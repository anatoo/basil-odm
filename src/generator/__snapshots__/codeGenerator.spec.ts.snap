// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateCollectionDefs() should works normally 1`] = `
"const $defs: Record<string, basil.EntityMeta<Record<string, unknown>>> = {
"foobar": {
  collectionName: "foobar",
  schema: new basil.FieldsSchema({
  "kind": "object",
  "props": {
    "_id": {
      "kind": "field",
      "isOptional": false,
      "node": {
        "kind": "objectId"
      }
    }
  },
  "allowAdditionalProps": false
}),
  indexes: [],
  options: {
  "comment": "foobar",
  "capped": true
},
},};

"
`;
